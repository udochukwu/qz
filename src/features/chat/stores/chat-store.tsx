import { SelectedNode } from '@/types';
import { create } from 'zustand';

interface chatStore {
  selectedNodes: SelectedNode[];
  addToSelectedNodes: (nodes: SelectedNode) => void;
  setSelectedNodes: (nodes: SelectedNode[]) => void;
}
const useChatStore = create<chatStore>(set => ({
  selectedNodes: [],
  // Add a new node to the selected nodes, if the message_id already exists,
  // it will be overwritten with the new selected_child
  addToSelectedNodes: (node: SelectedNode) =>
    set(state => ({
      // Filter out any node that has the same message_id as the new node
      selectedNodes: [
        ...state.selectedNodes.filter(n => n.message_id !== node.message_id),
        node, // Add the new node at the end of the array
      ],
    })),
  setSelectedNodes: (nodes: SelectedNode[]) => set(() => ({ selectedNodes: nodes })),
}));

export default useChatStore;
