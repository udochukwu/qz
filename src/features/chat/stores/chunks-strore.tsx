import { ResourceChunk, ResourceChunkWithMessageId } from '@/types';
import { create } from 'zustand';

interface ChunksStore {
  all_resource_chunks: { [message_id: string]: ResourceChunk[] };
  chunk_name_map: { [chunk_id: string]: number };
  lastChunkNumber: number;
  appendChunks: (message_id: string, resource_chunks: ResourceChunk[]) => void;
  resetChunksStore: () => void;
  getChunkName: (chunk_id: string) => number;
  getChunk: (chunk_id: string) => ResourceChunk[];
}

const useChunksStore = create<ChunksStore>((set, get) => ({
  all_resource_chunks: {},
  chunk_name_map: {},
  lastChunkNumber: 0,

  appendChunks: (message_id, resource_chunks) => {
    set(state => {
      const existingChunks = state.all_resource_chunks[message_id] || [];

      // Filter out duplicates and assign names to new chunks
      const newChunks = resource_chunks
        .filter(chunk => !existingChunks.some(existingChunk => existingChunk.chunk_id === chunk.chunk_id))
        .map(chunk => {
          // Only assign a new name if the chunk doesn't already have one
          if (!state.chunk_name_map[chunk.chunk_id]) {
            state.chunk_name_map[chunk.chunk_id] = state.lastChunkNumber + 1;
            state.lastChunkNumber += 1;
          }

          return {
            ...chunk,
            chunkName: state.chunk_name_map[chunk.chunk_id],
          };
        });

      if (!newChunks.length) return state;

      return {
        ...state,
        all_resource_chunks: {
          ...state.all_resource_chunks,
          [message_id]: [...existingChunks, ...newChunks],
        },
        chunk_name_map: state.chunk_name_map,
        lastChunkNumber: state.lastChunkNumber,
      };
    });
  },

  getChunkName: (chunk_id: string) => {
    return get().chunk_name_map[chunk_id] || 0;
  },

  getChunk: (chunk_id: string) => {
    const chunks = get().all_resource_chunks;
    for (const messageChunks of Object.values(chunks)) {
      const matchingChunk = messageChunks.find(chunk => chunk.chunk_id === chunk_id);
      if (matchingChunk) {
        return [matchingChunk];
      }
    }
    return [];
  },

  resetChunksStore: () => {
    set(() => ({
      all_resource_chunks: {},
      chunk_name_map: {},
      lastChunkNumber: 0,
    }));
  },
}));

export default useChunksStore;
