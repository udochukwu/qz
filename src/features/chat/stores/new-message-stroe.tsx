import { create } from 'zustand';
import { FeedbackData } from '../types';

interface NewMessageStore {
  messageFromRedirect: string | null;
  isNewMessageLoading: boolean; //When streaming
  feedbackData: FeedbackData | null;
  setMessageFromRedirect: (message: string | null) => void;
  setLoading: (isLoading: boolean) => void;
  setFeedbackData: (feedbackData: FeedbackData | null) => void;
}

const useNewMessageStore = create<NewMessageStore>(set => ({
  isNewMessageLoading: false,
  messageFromRedirect: null,
  feedbackData: null,
  setLoading: (isLoading: boolean) => set({ isNewMessageLoading: isLoading }),
  setMessageFromRedirect: (message: string | null) => set({ messageFromRedirect: message }),
  setFeedbackData: feedbackData => set({ feedbackData }),
}));

export default useNewMessageStore;
