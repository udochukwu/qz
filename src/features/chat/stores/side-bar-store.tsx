import { create } from 'zustand';

interface sideBarStore {
  isSideBarOpen: boolean;
  setSideBarOpen: (isOpen: boolean) => void;
}
const useSideBarStore = create<sideBarStore>(set => ({
  isSideBarOpen: true,
  setSideBarOpen: (isOpen: boolean) => set(() => ({ isSideBarOpen: isOpen })),
}));

export default useSideBarStore;
