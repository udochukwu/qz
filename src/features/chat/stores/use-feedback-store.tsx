import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { SHOW_FEEDBACK_FORM } from '../consts/feedback';

interface FeedbackStore {
  showFeedbackForm: boolean;
  setShowFeedbackForm: (show: boolean) => void;
}

const useFeedbackStore = create<FeedbackStore>()(
  persist(
    set => ({
      showFeedbackForm: false,
      setShowFeedbackForm: (show: boolean) => set({ showFeedbackForm: show }),
    }),
    { name: SHOW_FEEDBACK_FORM, version: 1 },
  ),
);

export default useFeedbackStore;
