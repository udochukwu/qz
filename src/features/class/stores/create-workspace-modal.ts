import { create } from 'zustand';

interface Store {
  isOpen?: boolean;
  setIsOpen: (val: boolean) => void;
}

export const useCreateWorkspaceModalStore = create<Store>(set => ({
  isOpen: false,
  setIsOpen: isOpen => set({ isOpen }),
}));
