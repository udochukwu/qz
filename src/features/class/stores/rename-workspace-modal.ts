import { create } from 'zustand';

interface Store {
  isOpen?: boolean;
  setIsOpen: (val: boolean) => void;
}

export const useRenameWorkspaceModalStore = create<Store>(set => ({
  isOpen: false,
  setIsOpen: isOpen => set({ isOpen }),
}));
