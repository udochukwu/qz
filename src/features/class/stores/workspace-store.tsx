import { create } from 'zustand';

interface WorkspaceStore {
  workspaceId: string | null;
  setWorkspaceId: (id: string) => void;
}

const useWorkspaceStore = create<WorkspaceStore>(set => ({
  workspaceId: null,
  setWorkspaceId: (id: string) => set(() => ({ workspaceId: id })),
}));

export default useWorkspaceStore;
