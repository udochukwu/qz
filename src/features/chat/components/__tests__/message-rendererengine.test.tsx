import { render, screen, fireEvent } from '@testing-library/react';
import MessageRendererV2 from '../message-rendererengine';
import { useUserStore } from '@/stores/user-store';
import { useTranslation } from 'react-i18next';

// Mock the dependencies
jest.mock('@/stores/user-store');
jest.mock('react-i18next');
jest.mock('react-markdown', () => {
  return function MockReactMarkdown({ children, className, components }: any) {
    // Process the content to handle custom components
    const content = typeof children === 'string' ? children : '';
    const parts = content.split(/(<mathpix[^>]*>.*?<\/mathpix>)/);

    return (
      <div data-testid="markdown-content" className={className}>
        {parts.map((part, index) => {
          const mathpixMatch = part.match(/<mathpix content="([^"]*)">/);
          if (mathpixMatch) {
            return <div key={index}>{components.mathpix({ content: mathpixMatch[1] })}</div>;
          }
          return <span key={index}>{part}</span>;
        })}
      </div>
    );
  };
});

// Mock styled-system components
jest.mock('@/components/elements/switch', () => ({
  Switch: ({ children, checked, onChange }: any) => (
    <button role="switch" aria-checked={checked} onClick={onChange}>
      {children}
    </button>
  ),
}));

// Mock ChunkButton
jest.mock('@/components/chunk-button', () => ({
  __esModule: true,
  default: ({ chunk_id }: { chunk_id: string }) => <button data-testid="chunk-button">{chunk_id}</button>,
}));

// Mock styled-system/css
jest.mock('styled-system/css', () => ({
  css: (...args: any[]) => args,
}));

// Mock rehype and remark plugins
jest.mock('rehype-katex', () => () => {});
jest.mock('rehype-raw', () => () => {});
jest.mock('remark-math', () => () => {});

jest.mock('mathpix-markdown-it', () => ({
  MathpixLoader: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  MathpixMarkdown: ({ text }: { text: string }) => <div data-testid="mathpix-content">{text}</div>,
}));

// Mock the latex processor
jest.mock('../../utils/latex-processor', () => ({
  processLatexContent: jest.fn((text: string) => {
    // Only process text that contains latex delimiters
    if (text.includes('$') || text.includes('\\')) {
      return `<mathpix content="${text}"></mathpix>`;
    }
    return text;
  }),
}));

// Mock MessageLoading and MessageErrorCard components
jest.mock('../message-loading', () => ({
  __esModule: true,
  default: () => <div data-testid="message-loading">Loading...</div>,
}));

jest.mock('../message-error-card', () => ({
  MessageErrorCard: () => <div data-testid="message-error-card">Error occurred</div>,
}));

describe('MessageRendererV2', () => {
  const mockT = jest.fn(key => key);
  const mockSetImpersonated = jest.fn();
  const mockUseUserStore = useUserStore as unknown as jest.Mock;

  beforeEach(() => {
    (useTranslation as jest.Mock).mockReturnValue({ t: mockT });
    mockUseUserStore.mockReturnValue({
      impersonated: false,
      setImpersonated: mockSetImpersonated,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render markdown content correctly', () => {
    const message = 'Hello **world**';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(screen.getByTestId('markdown-content')).toBeInTheDocument();
  });

  it('should render mathpix content correctly', () => {
    const message = '<mathpix content="E = mc^2"></mathpix>';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(screen.getByTestId('mathpix-content')).toBeInTheDocument();
    expect(screen.getByTestId('mathpix-content')).toHaveTextContent('E = mc^2');
  });

  it('should process latex content correctly', () => {
    const { processLatexContent } = require('../../utils/latex-processor');
    const message = 'The formula is $E = mc^2$';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(processLatexContent).toHaveBeenCalledWith(message);
    expect(processLatexContent).toHaveReturnedWith('<mathpix content="The formula is $E = mc^2$"></mathpix>');
  });

  it('should process multiple latex formulas in the same message', () => {
    const { processLatexContent } = require('../../utils/latex-processor');
    const message = 'First formula: $E = mc^2$ and second formula: $F = ma$';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(processLatexContent).toHaveBeenCalledTimes(1);
    expect(processLatexContent).toHaveBeenCalledWith(message);
    expect(processLatexContent).toHaveReturnedWith(
      '<mathpix content="First formula: $E = mc^2$ and second formula: $F = ma$"></mathpix>',
    );
  });

  it('should handle both inline and display latex formulas', () => {
    const { processLatexContent } = require('../../utils/latex-processor');
    const message = `
      Inline formula: $x^2 + y^2 = z^2$
      
      Display formula:
      $$
      \\int_{0}^{\\infty} e^{-x^2} dx = \\frac{\\sqrt{\\pi}}{2}
      $$
    `;
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(processLatexContent).toHaveBeenCalledTimes(1);
    expect(processLatexContent).toHaveBeenCalledWith(message);
    expect(processLatexContent).toHaveReturnedWith(`<mathpix content="${message}"></mathpix>`);
  });

  it('should not process non-latex content', () => {
    const { processLatexContent } = require('../../utils/latex-processor');
    const message = 'This is just regular text without any formulas';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(processLatexContent).toHaveBeenCalledWith(message);
    expect(processLatexContent).toHaveReturnedWith(message);
  });

  it('should handle escaped latex characters', () => {
    const { processLatexContent } = require('../../utils/latex-processor');
    const message = 'The formula is $\\alpha + \\beta = \\gamma$';
    render(<MessageRendererV2 message_id="1" message={message} />);

    expect(processLatexContent).toHaveBeenCalledWith(message);
    expect(processLatexContent).toHaveReturnedWith(
      '<mathpix content="The formula is $\\alpha + \\beta = \\gamma$"></mathpix>',
    );
  });

  // SNAPSHOT TESTS
  it('should match snapshot for markdown content', () => {
    const message = 'Hello **world** with some *italic* and `code`';
    const { container } = render(<MessageRendererV2 message_id="1" message={message} />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for mathpix content', () => {
    const message = '<mathpix content="E = mc^2"></mathpix>';
    const { container } = render(<MessageRendererV2 message_id="1" message={message} />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for mixed content', () => {
    const message = 'Hello **world** <mathpix content="E = mc^2"></mathpix> and some more text';
    const { container } = render(<MessageRendererV2 message_id="1" message={message} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle complex mathematical formulas correctly', () => {
    const complexMath = '<mathpix content="\\int_{0}^{\\infty} e^{-x^2} dx = \\frac{\\sqrt{\\pi}}{2}"></mathpix>';
    const { container } = render(<MessageRendererV2 message_id="1" message={complexMath} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle multiple mathpix formulas in sequence', () => {
    const multipleMath = `
      <mathpix content="\\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}"></mathpix>
      <mathpix content="\\sum_{i=1}^n i = \\frac{n(n+1)}{2}"></mathpix>
      <mathpix content="\\lim_{x \\to \\infty} \\frac{1}{x} = 0"></mathpix>
    `;
    const { container } = render(<MessageRendererV2 message_id="1" message={multipleMath} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle complex markdown with code blocks and inline code', () => {
    const complexMarkdown = `
      # Heading 1
      ## Heading 2
      
      This is a **bold** and *italic* text with \`inline code\`.
      
      \`\`\`python
      def hello_world():
          print("Hello, World!")
      \`\`\`
      
      - List item 1
      - List item 2
        - Nested item
      
      > Blockquote
      
      | Table | Header |
      |-------|---------|
      | Cell  | Cell    |
    `;
    const { container } = render(<MessageRendererV2 message_id="1" message={complexMarkdown} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle mixed content with complex markdown and math formulas', () => {
    const mixedComplexContent = `
      # Mathematical Analysis
      
      The quadratic formula is:
      <mathpix content="x = \\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}"></mathpix>
      
      ## Code Example
      \`\`\`python
      def quadratic_formula(a, b, c):
          discriminant = b**2 - 4*a*c
          return (-b + math.sqrt(discriminant))/(2*a)
      \`\`\`
      
      ### Important Notes
      - The **discriminant** determines the nature of the roots
      - When discriminant > 0: Two real roots
      - When discriminant = 0: One real root
      
      The integral of e^x is:
      <mathpix content="\\int e^x dx = e^x + C"></mathpix>
      
      > Note: C is the constant of integration
    `;
    const { container } = render(<MessageRendererV2 message_id="1" message={mixedComplexContent} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle special characters in mathpix content', () => {
    const specialChars = '<mathpix content="\\alpha + \\beta = \\gamma \\pm \\delta"></mathpix>';
    const { container } = render(<MessageRendererV2 message_id="1" message={specialChars} />);
    expect(container).toMatchSnapshot();
  });

  it('should handle markdown with links and images', () => {
    const markdownWithMedia = `
      [Link to documentation](https://example.com)
      
      ![Image description](https://example.com/image.jpg)
      
      <mathpix content="\\text{Image processing formula: } \\mathcal{F}\\{f(x,y)\\}"></mathpix>
    `;
    const { container } = render(<MessageRendererV2 message_id="1" message={markdownWithMedia} />);
    expect(container).toMatchSnapshot();
  });
});
