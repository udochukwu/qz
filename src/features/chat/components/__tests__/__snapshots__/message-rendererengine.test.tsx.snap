// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageRendererV2 should handle complex markdown with code blocks and inline code 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span>
        
      # Heading 1
      ## Heading 2
      
      This is a **bold** and *italic* text with \`inline code\`.
      
      \`\`\`python
      def hello_world():
          print("Hello, World!")
      \`\`\`
      
      - List item 1
      - List item 2
        - Nested item
      
      &gt; Blockquote
      
      | Table | Header |
      |-------|---------|
      | Cell  | Cell    |
    
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should handle complex mathematical formulas correctly 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\int_{0}^{\\infty} e^{-x^2} dx = \\frac{\\sqrt{\\pi}}{2}
          </div>
        </div>
      </div>
      <span>
        "&gt;&lt;/mathpix&gt;
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should handle markdown with links and images 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\text{Image processing formula: } \\mathcal{F}\\{f(x,y)\\}
          </div>
        </div>
      </div>
      <span>
        
    "&gt;&lt;/mathpix&gt;
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should handle mixed content with complex markdown and math formulas 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            x = \\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}
          </div>
        </div>
      </div>
      <span>
        
      
      ## Code Example
      \`\`\`python
      def quadratic_formula(a, b, c):
          discriminant = b**2 - 4*a*c
          return (-b + math.sqrt(discriminant))/(2*a)
      \`\`\`
      
      ### Important Notes
      - The **discriminant** determines the nature of the roots
      - When discriminant &gt; 0: Two real roots
      - When discriminant = 0: One real root
      
      The integral of e^x is:
      
      </span>
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\int e^x dx = e^x + C
          </div>
        </div>
      </div>
      <span>
        
      
      &gt; Note: C is the constant of integration
    "&gt;&lt;/mathpix&gt;
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should handle multiple mathpix formulas in sequence 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\frac{-b \\pm \\sqrt{b^2-4ac}}{2a}
          </div>
        </div>
      </div>
      <span>
        
      
      </span>
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\sum_{i=1}^n i = \\frac{n(n+1)}{2}
          </div>
        </div>
      </div>
      <span>
        
      
      </span>
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\lim_{x \\to \\infty} \\frac{1}{x} = 0
          </div>
        </div>
      </div>
      <span>
        
    "&gt;&lt;/mathpix&gt;
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should handle special characters in mathpix content 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            \\alpha + \\beta = \\gamma \\pm \\delta
          </div>
        </div>
      </div>
      <span>
        "&gt;&lt;/mathpix&gt;
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should match snapshot for markdown content 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span>
        Hello **world** with some *italic* and \`code\`
      </span>
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should match snapshot for mathpix content 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span />
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            E = mc^2
          </div>
        </div>
      </div>
      <span />
    </div>
  </article>
</div>
`;

exports[`MessageRendererV2 should match snapshot for mixed content 1`] = `
<div>
  <article
    class="message-renderer"
  >
    <div
      data-testid="markdown-content"
    >
      <span>
        Hello **world** 
      </span>
      <div>
        <div>
          <div
            data-testid="mathpix-content"
          >
            E = mc^2
          </div>
        </div>
      </div>
      <span>
         and some more text
      </span>
    </div>
  </article>
</div>
`;
